### PR TITLE
Stock watchlist improvements

### PR DESCRIPTION
**Pull request by kamalq97**
Changelog:
- Combined `watchlist_add` and `watchlist_remove` into a new url/view called `watchlist_edit` in **urls.py** and **views.py**.
- New method for error handling `watchlist_edit_view` when stock symbol is not found (similar to how errors are handled in the `single_stock` view)
- Added `getCookie` function to **base.html** (mainly used for fetching the "csrftoken" cookie needed for post requests)
- Added JavaScript code to **index.html**, **single_stock.htm**l, and **watchlist.html**. The code executes on the click of the blue or red watchlist button.

 _Now, adding/removing a stock from the watchlist no longer forces the entire webpage to reload._

### CODE REVIEW

File: myapp/templates/index.html [hunk 1]
```diff
@@ -33,21 +33,38 @@
...
             </tr>
         {% endfor %}
     </table>
-{% endblock %}
\ No newline at end of file
+{% endblock %}
+
+{% block scripts %}
+<script>
+    function watchlistEdit(symbol, number) {
```
Comment (murradkh) [inline myapp/templates/index.html:102]: Nice Kamal, you learned AJAX very quickly!

File: myapp/templates/index.html [hunk 2]
```diff
@@ -33,21 +33,38 @@
...
+            xhr.open("POST", `stock/${symbol}/wadd/`, true);
+            icon.className = "fa fa-eye-slash";
+            editButton.className = "btn btn-outline-danger";
+        } 
+        else {
+            xhr.open("POST", `stock/${symbol}/wremove/`, true);
+            icon.className = "fa fa-eye";
+            editButton.className = "btn btn-outline-primary";     
+        }
+        xhr.setRequestHeader('X-CSRFToken', getCookie('csrftoken'));
```
Comment (murradkh) [inline myapp/templates/index.html:132]: i think you don't need for `getCookie` method to fetch the cookie, you can use `xhr.withCredentials = true;` to include the cookie into the post request
reference: [URL1](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/withCredentials), [URL2](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS)


let me know if you stuck with it

File: myapp/templates/index.html [hunk 3]
```diff
@@ -33,21 +33,38 @@
...
+        xhr.send();
```
Comment (murradkh) [inline myapp/templates/index.html:133]: what would happen if the request failed for some reason? is the icon would be changed anyway? 
preferred to have a callback function, to be called once the response received from the server.

File: myapp/templates/single_stock.html [hunk 4]
```diff
@@ -124,6 +118,23 @@
...
+            xhr.open("POST", "{% url 'watchlist_edit' symbol=data.symbol operation='wadd' %}", true);
+            editButton.innerHTML = "<i class='fa fa-eye-slash'></i> Remove from watchlist";
+            editButton.className = "btn btn-outline-danger";   
+        } 
+        else {
+            xhr.open("POST", "{% url 'watchlist_edit' symbol=data.symbol operation='wremove' %}", true);
+            editButton.innerHTML = "<i class='fa fa-eye'></i> Add to watchlist";
+            editButton.className = "btn btn-outline-primary";            
+        }
+        xhr.setRequestHeader('X-CSRFToken', getCookie('csrftoken'));
```
Comment (murradkh) [inline myapp/templates/single_stock.html:152]: the same

File: myapp/urls.py [hunk 5]
```diff
@@ -10,8 +10,7 @@
 	path('accounts/login/', auth_views.LoginView.as_view(template_name='login.html'), name='login'),
 	path('accounts/logout/', views.logout_view, name='logout'),
 	path('accounts/watchlist/', views.watchlist_view, name='watchlist'),
-	path('stock/<str:symbol>/wremove/', views.watchlist_remove_view, name='watchlist_remove'),
-	path('stock/<str:symbol>/wadd/', views.watchlist_add_view, name='watchlist_add'),
+	path('stock/<str:symbol>/<str:operation>/', views.watchlist_edit_view, name='watchlist_edit'),
```
Comment (murradkh) [inline myapp/urls.py]: Nice!

File: myapp/views.py [hunk 6]
```diff
@@ -88,28 +88,20 @@ def watchlist_view(request):
...
     else:
-        Stock.remove_from_watchlist(profile, symbol)
-        next = request.POST.get('next', '/')
-        return redirect(next)
+        if operation == "wremove":
+            Stock.remove_from_watchlist(profile, symbol)
+        if operation == "wadd":
+            Stock.add_to_watchlist(profile, symbol)
+        nextpage = request.POST.get('next', '/')
+        return redirect(nextpage)               # NOTE: last two lines no longer needed, remove?
```
Comment (murradkh) [inline myapp/views.py]: I'm not sure why "next" is needed and where is assigned to the value

**Review by murradkh — COMMENTED** [review]
Good Job, many things covered

File: myapp/templates/index.html (hunk 1 above)
Comment (kamalq97) [reply myapp/templates/index.html:102]: Thanks!

File: myapp/templates/index.html (hunk 3 above)
Comment (kamalq97) [reply myapp/templates/index.html:133]: Good point!

File: myapp/views.py (hunk 6 above)
Comment (kamaql97) [reply myapp/views.py:104]: It is no longer needed, lines 103 and 104 can be safely deleted

File: myapp/templates/index.html (hunk 2 above)
Comment (kamaql97) [reply myapp/templates/index.html:66]: I will give it a try

Comment (kamalq97) [reply myapp/templates/index.html:132]: Unfortunately, the method you talk about did not work, it gave the following error message: 
`Forbidden (CSRF token missing or incorrect.)`

Comment (murradkh) [reply myapp/templates/index.html:132]: i will take a look

File: myapp/templates/index.html (hunk 3 above)
Comment (kamalq97) [reply myapp/templates/index.html:133]: @murradkh Fixed! Check my newest commit

File: myapp/templates/single_stock.html (hunk 4 above)
Comment (kamalq97) [reply myapp/templates/single_stock.html:152]: @murradkh Fixed! Check my newest commit

File: myapp/views.py (hunk 6 above)
Comment (kamalq97) [reply myapp/views.py]: It is no longer needed, we can return some other response

File: myapp/templates/index.html (hunk 2 above)
Comment (murradkh) [reply myapp/templates/index.html:132]: looks like there is no shortage/alternative way other than what you did to render the `csrfmiddlewaretoken` value
good to go with this.

File: myapp/templates/single_stock.html (hunk 4 above)
Comment (murradkh) [reply myapp/templates/single_stock.html:152]: okay

File: myapp/views.py (hunk 6 above)
Comment (murradkh) [reply myapp/views.py]: you can change to `return HttpResponse("OK")`  (redirecting cause to create redundant request)

**Review by murradkh — APPROVED** [review]
looks better! 
good to merge after you resolve the conflicts

**Comment by kamalq97** [comment]
Yes, I will resolve the conflicts 👍

**Pull request merged**